tests/monitoring/monitoring_service/test_crash.py

```diff
@@ -0,0 +1,125 @@
+        from_block = chain_state.latest_known_block + 1
+        blocks = self.block_events[from_block:to_block + 1]
+        events = [ev for block in blocks for ev in block]  # flatten
+        return chain_state, events
+
+
+def test_crash(
+    monitoring_service: MonitoringService,  # adds stake in ServiceRegistry
+    web3,
+    contracts_manager,
+    server_private_key,
+    token_network_registry_contract,
+    monitoring_service_contract,
+    get_random_address,
+    tmpdir,
+    generate_raiden_clients,
+    token_network,
+):
+    """ Process blocks and compare results with/without crash
+
+    A somewhat meaninful crash handling is simulated by not including the
+    UpdatedHeadBlockEvent in every block.
+    """
+    channel_identifier = 3
+    c1, c2 = generate_raiden_clients(2)
+    monitor_request = c2.get_monitor_request(
+        balance_proof=c1.get_balance_proof(
+            channel_id=channel_identifier,
+            nonce=1,
+            additional_hash='0x11',
+            transferred_amount=2,
+            locked_amount=0,
+            locksroot='0x00',
+        ),
+        reward_amount=0,
+    )
+
+    events = [
+        [
+            ReceiveChannelOpenedEvent(
+                token_network.address,
+                channel_identifier,
+                c1.address, c2.address,
+                settle_timeout=20,
+                block_number=0),
+        ],
+        [
+            UpdatedHeadBlockEvent(1),
+        ],
+        [
+            ActionMonitoringTriggeredEvent(
+                token_network.address,
+                channel_identifier,
+                non_closing_participant=c2.address,
+            ),
+        ],
+        [
+            UpdatedHeadBlockEvent(3),
+        ],
+    ]
+
+    def new_ms(filename):
+        ms = MonitoringService(
+            web3=web3,
+            contract_manager=contracts_manager,
+            private_key=server_private_key,
+            registry_address=token_network_registry_contract.address,
+            monitor_contract_address=monitoring_service_contract.address,
+            db_filename=os.path.join(tmpdir, filename),
+        )
+        ms.bcl = MockBlockchainListener(events)  # type: ignore
+        msc = Mock()
+        ms.context.monitoring_service_contract = msc
+        ms.monitor_mock = msc.functions.monitor.return_value.transact  # type:ignore
+        ms.monitor_mock.return_value = bytes(0)  # type:ignore
+        return ms
+
+    # initialize both monitoring services
+    stable_ms = new_ms('stable.db')
+    crashy_ms = new_ms('crashy.db')
+    for ms in [stable_ms, crashy_ms]:
+        ms.database.conn.execute(
+            "INSERT INTO token_network(address) VALUES (?)",
+            [token_network.address],
+        )
+        ms.context.ms_state.blockchain_state.token_network_addresses = [token_network.address]
+        ms.database.upsert_monitor_request(monitor_request)
+        ms.database.conn.commit()
```

Comment:
General question: Do we have to call commit after every db change, or at least in the `updated_head_block_event_handler`?

Reply:
With our current settings, all changes which are not explicitly committed are lost on crash/shutdown. The best way to handle a transaction is to use the connection as a context manager, which automatically commits on successful exit and rolls back when an exception is thrown.

As long as we're working synchronously, we could also choose to wrap larger sections of code into a transaction to build larger sets of atomic changes.